### PR TITLE
Adding instrcutions for SageMaker RL

### PR DESCRIPTION
**Pull request by yoheigon**
### Proposed change(s)

Describe the changes made in this PR.

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [x] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by yoheigon** [comment]
@chriselion 
Could you review this PR?

**Comment by chriselion** [comment]
Sorry, we need to discuss this internally more, and some of the relevant people are on vacation this week.

**Comment by unityjeffrey** [comment]
hi @yoheigon hope you are doing well.  Apologies for the late reply.  Thank you for writing this up, this would be beneficial for our users who are wanting to use ML-Agents with Sagemaker.  I was wondering if this document would be better suited to be hosted on the Amazon site (and we can link from here).  We have gotten away from providing documentation and support to run ml-agents on cloud providers as there is a support burden and knowledge gap on our end.

if you think it makes sense to include it in the core ml-agents documentation, then I would suggest the following changes:

1.  In the title, please include a note that this is only tested for a specific version of ML-Agents.  This will prevent users from trying to follow these instructions on a later version of ML-Agents.
2.  Please also include a note to indicate that this isn't generally supported by the ML-Agents team, but that it's included in case it may be helpful
3.  If folks have any questions or additional followups, is there an Amazon POC or support email address that can be added?  We end up getting questions in the forums or issues and would be good to be able to route them to someone at Amazon.

For 1-2, the Azure doc is a good example.  https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-on-Microsoft-Azure.md

Let me know on my note above, I'll also make some comments in the documentation itself.

Thanks!

Jeff

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
+## Training on Amazon SageMaker RL
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:1]: please use #

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+## Overview of architecture
+
+SageMaker is a fully managed service that enables fast model development. It provides many built-in features to assist you with training, tuning, debugging, and model deployment. SageMaker RL builds on top of SageMaker, adding pre-built RL libraries and making it easy to integrate with different simulation environments. You can use built-in deep learning frameworks such as TensorFlow and PyTorch with various built-in RL algorithms from the [RLlib](https://docs.ray.io/en/latest/rllib.html) library to train RL policies. Infrastructures for training and inference are fully managed by SageMaker, so you can focus on RL formulation. SageMaker RL also provides a set of [Jupyter notebooks](https://github.com/aws/amazon-sagemaker-examples/tree/master/reinforcement_learning), demonstrating varieties of domain RL applications in robotics, operations research, finance, and more.
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:7]: this section is a bit too much marketing - can you instead create a link to Sagemaker RL instead?  The architecture diagram below is helpful though.

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+The following diagram illustrates our architecture.
+![Architecture](images/architecture_sagemaker.png)
+
+## Configuring your environment
+
+In this section, we walk you through the steps to set up your environment to launch SageMaker RL training jobs.
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:13]: "we **will** walk"

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+### Setting up your environments
+
+To get started, we import the needed Python libraries and set up environments for permissions and configurations. The following code contains the steps to set up an [Amazon Simple Storage Service](http://aws.amazon.com/s3) (Amazon S3) bucket, define the training job prefix, specify the training job location, and create an [AWS Identity and Access Management](http://aws.amazon.com/iam) (IAM) role:
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:17]: "we **first need to** import"

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+1. In the Unity Editor, load a project containing an ML-Agents environment (you can use one of the example environments if you have not created your own but please make sure your environment only contains one single agent).
+
+2. Open the Build Settings window (menu: File > Build Settings).
+3. Select Linux as the Target Platform, and x86_64 as the target architecture.
+4. Click Build to build the Unity environment executable.
+5. Upload the executable file, dependency data files and library files to s3.
+
+## Model training, evaluation, and deployment
+
+In this section, we walk you through the steps to train, evaluate, and deploy models.
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:99]: "we **will** walk"

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+At a high level, the training job initiates a neural network and updates the network gradually towards the direction in which the agent collects higher reward. Through multiple trials, the agent eventually learns how to navigate to the high-rewarding location efficiently. SageMaker RL handles the entire process and allows you to view the training job status in the Training jobs page on the SageMaker console.
+
+### Evaluating the model
+
+When model training is complete, we can load the trained model to evaluate its performance. Similar to the setup in the training script, we wrap the Unity environment with a gym wrapper. We then create an agent by loading the trained model.
+To evaluate the model, we run the trained agent multiple times against the environment with a fixed agent and target initializations, and add up the cumulative rewards the agent collects at each step for each episode.
+Out of five episodes, the average episode reward is 0.92 with the maximum reward of 0.93 and minimum reward of 0.89, suggesting the trained model indeed performs well.
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:246]: this seems out of place.  I would put a qualifier like "For example, out of the five episodes..."

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+# Test to verify if all required dependencies installed successfully or not.
+RUN python -c "import gym;import sagemaker_containers.cli.train;import ray; from sagemaker_containers.cli.train import main; from mlagents_envs.environment import UnityEnvironment; from mlagents_envs.registry import default_registry; from gym_unity.envs import UnityToGymWrapper"
+
+# Make things a bit easier to debug
+WORKDIR /opt/ml/code
+```
+
+### Preparing your Unity environment
+
+1. In the Unity Editor, load a project containing an ML-Agents environment (you can use one of the example environments if you have not created your own but please make sure your environment only contains one single agent).
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:90]: I would make a note that this example uses the Basic environment and link it (https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Learning-Environment-Examples.md#basic)

File: docs/Training-on-Amazon-SageMaker-RL.md [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+By default, RLLib applies a pre-defined [convolutional neural network](https://github.com/ray-project/ray/blob/master/rllib/models/tf/visionnet.py) or [fully connected neural network](https://github.com/ray-project/ray/blob/master/rllib/models/tf/fcnet.py). However, you can create a custom model for training and testing. Following the [examples](https://github.com/ray-project/ray/blob/master/rllib/examples/custom_env.py) from RLLib, you can register the custom model by calling `ModelCatalog.register_custom_model`, then refer to the newly registered model using the `custom_model` argument.
+In our code example, we invoke a predefined Unity environment called Basic, but you can experiment with other [pre-built Unity environments](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Learning-Environment-Examples.md). However, as of this writing, our solution only supports a single-agent environment. When new environments are built, register it by calling `register_env` and refer to the environment with the env parameter.
```
Comment (unityjeffrey) [inline docs/Training-on-Amazon-SageMaker-RL.md:283]: I would put this note around single-agent environment as a separate disclaimer or move it up in the document.

**Review by unityjeffrey — CHANGES_REQUESTED** [review]

**Pull request closed without merging**